### PR TITLE
[apollo-normalized-cache] Convert remaining Java classes to Kotlin

### PR DESCRIPTION
**Pull request by martinbonnin**
This is a plain Java -> Kotlin conversion. There should be (almost) no behaviour change (see comments in the diff for more details).

Next steps:

* use the Kotlin `multiplatform` plugin
* move `ApolloStore` and `ApolloStoreOperation` to `jvmMain`. These two classes are based on `Executors` and `AtomicReferences` so they will be hard to convert to KMP. I think we can use another API  for KMP, based on coroutines.
* move `LruNormalizedCache` to `jvmMain` as well as it depends on `com.nytimes.android:cache`, which is not KMP
* when `com.dropbox.mobile.store:cache4` becomes KMP (https://github.com/apollographql/apollo-android/issues/2331#issuecomment-658621856), move everything to `commonMain`, potentially in another module?

### CODE REVIEW

File: apollo-normalized-cache/src/main/java/com/apollographql/apollo/cache/normalized/internal/NoOpApolloStore.kt [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+  override fun <R> readTransaction(transaction: Transaction<ReadableStore, R>): R {
+    return transaction.execute(this)!!
+  }
+
+  override fun <R> writeTransaction(transaction: Transaction<WriteableStore, R>): R {
+    return transaction.execute(this)!!
+  }
+
+  override fun normalizedCache(): NormalizedCache {
+    error("Cannot get normalizedCache: no cache configured")
```
Comment (martinbonnin) [inline apollo-normalized-cache/src/main/java/com/apollographql/apollo/cache/normalized/internal/NoOpApolloStore.kt:74]: This is a behaviour change. It used to return `null` here, which I believe would have crashed too.

File: apollo-normalized-cache/src/main/java/com/apollographql/apollo/cache/normalized/internal/NoOpApolloStore.kt [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+  override fun <D : Operation.Data, T, V : Operation.Variables> read(
+      operation: Operation<D, T, V>): ApolloStoreOperation<T> {
+    error("Cannot read operation: no cache configured")
+  }
+
+  override fun <D : Operation.Data, T, V : Operation.Variables> read(
+      operation: Operation<D, T, V>, responseFieldMapper: ResponseFieldMapper<D>,
+      responseNormalizer: ResponseNormalizer<Record>, cacheHeaders: CacheHeaders): ApolloStoreOperation<Response<T>> {
+    error("Cannot read operation: no cache configured")
```
Comment (martinbonnin) [inline apollo-normalized-cache/src/main/java/com/apollographql/apollo/cache/normalized/internal/NoOpApolloStore.kt:89]: Same as above, behaviour change. It used to return emptyOperation(null), which I believe would also crash. The rationale here is that calling these do not make any sense if no cache is configured.

File: apollo-normalized-cache/src/main/java/com/apollographql/apollo/cache/normalized/internal/NoOpApolloStore.kt [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    error("Cannot write fragment: no cache configured")
+  }
+
+  override fun writeAndPublish(fragment: GraphqlFragment, cacheKey: CacheKey,
+                               variables: Operation.Variables): ApolloStoreOperation<Boolean> {
+    error("Cannot writeAndPublish fragment: no cache configured")
+  }
+
+  override fun <D : Operation.Data, T, V : Operation.Variables> writeOptimisticUpdates(operation: Operation<D, T, V>, operationData: D, mutationId: UUID): ApolloStoreOperation<Set<String>> {
+    return emptyOperation(emptySet())
```
Comment (martinbonnin) [inline apollo-normalized-cache/src/main/java/com/apollographql/apollo/cache/normalized/internal/NoOpApolloStore.kt:118]: For the optimistic methods, I kept `emptyOperation` as I was unsure if they could be called.

File: apollo-normalized-cache/src/main/java/com/apollographql/apollo/cache/normalized/internal/NoOpApolloStore.kt (hunk 2 above)
Comment (martinbonnin) [reply apollo-normalized-cache/src/main/java/com/apollographql/apollo/cache/normalized/internal/NoOpApolloStore.kt:89]: Ok, that didn't work. The NO_OP cache is always called, looking into it...

Comment (martinbonnin) [reply apollo-normalized-cache/src/main/java/com/apollographql/apollo/cache/normalized/internal/NoOpApolloStore.kt]: Looks like only this method was called. I still made the other calls safer when it was possible.

File: apollo-normalized-cache/src/main/java/com/apollographql/apollo/cache/normalized/internal/SimpleStack.kt [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package com.apollographql.apollo.cache.normalized.internal
+
+import java.util.ArrayList
+
+/**
+ * Simple stack data structure which accepts null elements. Backed by list.
+ * @param <E>
+</E> */
+class SimpleStack<E> {
+  private var backing: MutableList<E>
```
Comment (sav007) [inline apollo-normalized-cache/src/main/java/com/apollographql/apollo/cache/normalized/internal/SimpleStack.kt:10]: nit: I guess can be `val`?

**Comment by martinbonnin** [comment]
Superseded by https://github.com/apollographql/apollo-android/pull/2457 (same commits but using a local branch and not from my fork so that I could merge the dependant PR first)

**Pull request closed without merging**